v1/coordinator/statement_executor.go

```diff
@@ -385,53 +386,125 @@ func (e *StatementExecutor) executeDropMeasurementStatement(ctx context.Context,
 	return e.TSDBStore.DeleteMeasurement(ctx, mapping.BucketID.String(), q.Name)
 }
 
+type measurementRow struct {
+	name   []byte
+	db, rp string
+}
+
 func (e *StatementExecutor) executeShowMeasurementsStatement(ctx context.Context, q *influxql.ShowMeasurementsStatement, ectx *query.ExecutionContext) error {
-	if q.Database == "" {
+	if q.Database == "" && !q.WildcardDatabase {
 		return ErrDatabaseNameRequired
 	}
 
-	mapping, err := e.getDefaultRP(ctx, q.Database, ectx)
+	if q.WildcardDatabase {
+		// We could support this but it doesn't seem very useful.
+		if q.RetentionPolicy != "" {
+			return ectx.Send(ctx, &query.Result{
+				Err: fmt.Errorf("query 'SHOW MEASUREMENTS ON *.rp' not supported. use 'ON *.*' or specify a database"),
+			})
+		}
+		// It is not clear how '*' should interact with the default retention policy, so reject it
+		if !q.WildcardRetentionPolicy {
+			return ectx.Send(ctx, &query.Result{
+				Err: fmt.Errorf("query 'SHOW MEASUREMENTS ON *' not supported. use 'ON *.*' or specify a database"),
+			})
+		}
+	}
+
+	onlyPrintMeasurements := !(q.WildcardDatabase || q.WildcardRetentionPolicy || q.RetentionPolicy != "")
+
+	mappingsFilter := influxdb.DBRPMappingFilter{
+		OrgID: &ectx.OrgID,
+	}
+
+	if !q.WildcardDatabase {
+		mappingsFilter.Database = &q.Database
+	}
+	if !q.WildcardRetentionPolicy {
+		if q.RetentionPolicy == "" {
+			defaultRP := true
+			mappingsFilter.Default = &defaultRP
+		} else {
+			mappingsFilter.RetentionPolicy = &q.RetentionPolicy
+		}
+	}
+	mappings, _, err := e.DBRP.FindMany(ctx, mappingsFilter)
 	if err != nil {
-		return err
+		return fmt.Errorf("finding DBRP mappings: %v", err)
 	}
 
-	names, err := e.TSDBStore.MeasurementNames(ctx, ectx.Authorizer, mapping.BucketID.String(), q.Condition)
-	if err != nil || len(names) == 0 {
-		return ectx.Send(ctx, &query.Result{
-			Err: err,
-		})
+	rows := make([]measurementRow, 0)
+
+	// Sort the sources for consistent output
+	sort.Slice(mappings, func(i, j int) bool {
+		if mappings[i].Database != mappings[j].Database {
+			return mappings[i].Database < mappings[j].Database
+		}
+		return mappings[i].RetentionPolicy < mappings[j].RetentionPolicy
+	})
+
+	for _, mapping := range mappings {
+		names, err := e.TSDBStore.MeasurementNames(ctx, ectx.Authorizer, mapping.BucketID.String(), q.Condition)
+		if err != nil || len(names) == 0 {
+			return ectx.Send(ctx, &query.Result{
+				Err: err,
```

Comment:
Should we separate these cases, `err != nil` and `len(names) == 0` and make up an error for the `len() == 0`?  Or are you counting on the empty query result with a nil error as being interpreted as no error but no data? 

If the latter, would it be cleaner to fall through and have a single exit point for no results (whether by initial return, or by limit and offset?

Reply:
Drat this is a bug - will fix

Reply:
@davidby-influx great catch, thanks!